File: scripts/evaluate_detection.py
#!/usr/bin/env python3
"""Evaluate the CLI detector against the shared annotated fixture set."""

from __future__ import annotations

import argparse
import json
import math
import statistics
import sys
import time
from pathlib import Path

import numpy as np
from PIL import Image, ImageDraw, ImageOps


REPO_ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(REPO_ROOT / "src"))

from slides_thief.cli import detect_quad  # noqa: E402


FEATURE_NAME_MAP = {
    "edge_strength": "edgeStrength",
    "edge_support": "edgeSupport",
    "edge_continuity": "edgeContinuity",
    "gradient_alignment": "gradientAlignment",
    "inside_outside_difference": "insideOutsideDifference",
    "region_consistency": "regionConsistency",
    "normalized_area": "normalizedArea",
    "geometry_validity": "geometryValidity",
    "aspect_prior": "aspectPrior",
    "batch_consistency": "batchConsistency",
}


def canonical_features(features: dict) -> dict[str, float]:
    return {FEATURE_NAME_MAP[name]: float(value) for name, value in features.items()}


def quad_iou(predicted: np.ndarray, expected: np.ndarray, width: int, height: int) -> float:
    predicted_mask = Image.new("1", (width, height))
    expected_mask = Image.new("1", (width, height))
    ImageDraw.Draw(predicted_mask).polygon([tuple(point) for point in predicted], fill=1)
    ImageDraw.Draw(expected_mask).polygon([tuple(point) for point in expected], fill=1)
    predicted_arr = np.asarray(predicted_mask, dtype=bool)
    expected_arr = np.asarray(expected_mask, dtype=bool)
    union = np.logical_or(predicted_arr, expected_arr).sum()
    return float(np.logical_and(predicted_arr, expected_arr).sum() / union) if union else 0.0


def percentile(values: list[float], fraction: float) -> float:
    if not values:
        return 0.0
    ordered = sorted(values)
    index = min(len(ordered) - 1, max(0, math.ceil(len(ordered) * fraction) - 1))
    return ordered[index]


def evaluate(annotations_path: Path) -> tuple[dict, dict]:
    fixture_root = annotations_path.parent
    annotations = json.loads(annotations_path.read_text(encoding="utf-8"))
    rows = []
    runtimes = []

    for item in annotations["images"]:
        image_path = fixture_root / item["file"]
        image = ImageOps.exif_transpose(Image.open(image_path)).convert("RGB")
        started = time.perf_counter()
        ratio = float(item.get("ratio", 16 / 9))
        batch_priors = [
            {
                **prior,
                "normalized_quad": np.asarray(prior["normalized_quad"], dtype=np.float64),
            }
            for prior in item.get("batch_priors", [])
        ]
        predicted, diagnostics = detect_quad(
            image,
            ratio,
            batch_priors=batch_priors,
            enable_batch_prior=bool(batch_priors),
        )
        runtimes.append((time.perf_counter() - started) * 1000)
        expected = item.get("quad")
        if expected is None:
            errors = None
            iou = None
        else:
            expected = np.asarray(expected, dtype=np.float64)
            diagonal = math.hypot(image.width, image.height)
            errors = np.linalg.norm(predicted - expected, axis=1) / diagonal
            iou = quad_iou(predicted, expected, image.width, image.height)
        rows.append(
            {
                "file": item["file"],
                "quad": [[round(float(x), 4), round(float(y), 4)] for x, y in predicted],
                "method": diagnostics["method"],
                "confidence": diagnostics["confidence"],
                "needs_review": diagnostics["needs_review"],
                "review_reasons": diagnostics["review_reasons"],
                "candidate_methods": [
                    candidate["method"]
                    for candidate in diagnostics["diagnostics"].get("ranked_candidates", [])
                ],
                "candidate_count": diagnostics["diagnostics"].get("candidate_count_after_validation", 0),
                "best_score": diagnostics["best_score"],
                "second_best_score": diagnostics["second_best_score"],
                "selectedFeatures": canonical_features(diagnostics.get("features", {})),
                "selected_warnings": diagnostics["diagnostics"].get("selected_warnings", []),
                "confidence_breakdown": diagnostics["diagnostics"].get("confidence_breakdown"),
                "mean_corner_error": float(errors.mean()) if errors is not None else None,
                "max_corner_error": float(errors.max()) if errors is not None else None,
                "quad_iou": iou,
            }
        )

    accuracy_rows = [row for row in rows if row["max_corner_error"] is not None]
    mean_errors = [row["mean_corner_error"] for row in accuracy_rows]
    metrics = {
        "fixture_count": len(rows),
        "mean_corner_error": statistics.fmean(mean_errors) if mean_errors else 0.0,
        "median_corner_error": statistics.median(mean_errors) if mean_errors else 0.0,
        "max_corner_error_mean": statistics.fmean(row["max_corner_error"] for row in accuracy_rows)
        if accuracy_rows else 0.0,
        "quad_iou_mean": statistics.fmean(row["quad_iou"] for row in accuracy_rows) if accuracy_rows else 0.0,
        "all_corners_under_1_percent": (
            sum(row["max_corner_error"] < 0.01 for row in accuracy_rows) / len(accuracy_rows)
            if accuracy_rows else 0.0
        ),
        "all_corners_under_2_percent": (
            sum(row["max_corner_error"] < 0.02 for row in accuracy_rows) / len(accuracy_rows)
            if accuracy_rows else 0.0
        ),
        "review_rate": sum(row["needs_review"] for row in rows) / len(rows) if rows else 0.0,
        "high_confidence_failure_rate": (
            sum(row["confidence"] >= 0.8 and row["quad_iou"] < 0.75 for row in accuracy_rows)
            / len(accuracy_rows)
            if accuracy_rows else 0.0
        ),
        "runtime_p50_ms": statistics.median(runtimes) if runtimes else 0.0,
        "runtime_p95_ms": percentile(runtimes, 0.95),
    }
    predictions = {"images": rows}
    return metrics, predictions


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--annotations",
        type=Path,
        default=REPO_ROOT / "tests" / "fixtures" / "detection" / "annotations.json",
    )
    parser.add_argument("--output", type=Path)
    parser.add_argument("--predictions-output", type=Path)
    args = parser.parse_args()
    metrics, predictions = evaluate(args.annotations)
    payload = json.dumps(metrics, indent=2, sort_keys=True)
    if args.output:
        args.output.write_text(f"{payload}\n", encoding="utf-8")
    if args.predictions_output:
        args.predictions_output.write_text(
            f"{json.dumps(predictions, indent=2, sort_keys=True)}\n",
            encoding="utf-8",
        )
    print(payload)


if __name__ == "__main__":
    main()

File: site/scripts/evaluate-detection.mjs
#!/usr/bin/env node

import { readFile, writeFile } from "node:fs/promises";
import { performance } from "node:perf_hooks";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

import { detectQuad } from "../app/detection/detect.ts";
import { quadIoU } from "../app/detection/geometry.ts";

const siteRoot = resolve(dirname(fileURLToPath(import.meta.url)), "..");
const repoRoot = resolve(siteRoot, "..");
const args = process.argv.slice(2);
const outputFlag = args.indexOf("--output");
const outputPath = outputFlag >= 0 ? resolve(args[outputFlag + 1]) : null;
const predictionsFlag = args.indexOf("--predictions-output");
const predictionsPath = predictionsFlag >= 0 ? resolve(args[predictionsFlag + 1]) : null;
const positional = args.filter((value, index) =>
  !value.startsWith("--") && index !== outputFlag + 1 && index !== predictionsFlag + 1
);
const annotationsPath = resolve(
  positional[0] ?? join(repoRoot, "tests/fixtures/detection/annotations.json"),
);

function parsePpm(buffer) {
  let offset = 0;
  const token = () => {
    while (offset < buffer.length && /\s/.test(String.fromCharCode(buffer[offset]))) offset += 1;
    if (buffer[offset] === 35) {
      while (offset < buffer.length && buffer[offset] !== 10) offset += 1;
      return token();
    }
    const start = offset;
    while (offset < buffer.length && !/\s/.test(String.fromCharCode(buffer[offset]))) offset += 1;
    return buffer.subarray(start, offset).toString("ascii");
  };
  if (token() !== "P6") throw new Error("Only binary P6 PPM fixtures are supported.");
  const width = Number(token());
  const height = Number(token());
  if (Number(token()) !== 255) throw new Error("Only 8-bit PPM fixtures are supported.");
  while (offset < buffer.length && /\s/.test(String.fromCharCode(buffer[offset]))) offset += 1;
  const rgb = buffer.subarray(offset);
  const rgba = new Uint8ClampedArray(width * height * 4);
  for (let source = 0, target = 0; source < rgb.length; source += 3, target += 4) {
    rgba[target] = rgb[source];
    rgba[target + 1] = rgb[source + 1];
    rgba[target + 2] = rgb[source + 2];
    rgba[target + 3] = 255;
  }
  return { width, height, data: rgba };
}

function percentile(values, fraction) {
  if (!values.length) return 0;
  const ordered = [...values].sort((a, b) => a - b);
  return ordered[Math.min(ordered.length - 1, Math.max(0, Math.ceil(ordered.length * fraction) - 1))];
}

const annotations = JSON.parse(await readFile(annotationsPath, "utf8"));
const fixtureRoot = dirname(annotationsPath);
const rows = [];
const runtimes = [];

for (const item of annotations.images) {
  const ppmFile = item.file.replace(/\.[^.]+$/, ".ppm");
  const image = parsePpm(await readFile(join(fixtureRoot, ppmFile)));
  const started = performance.now();
  const result = detectQuad(image, {
    maxDetectionWidth: 900,
    sourceRatioHint: item.ratio ?? 16 / 9,
    enableBatchPrior: Boolean(item.batch_priors?.length),
  }, item.batch_priors ?? []);
  runtimes.push(performance.now() - started);
  const errors = item.quad
    ? result.quad.map((point, index) =>
      Math.hypot(point[0] - item.quad[index][0], point[1] - item.quad[index][1]) /
      Math.hypot(image.width, image.height)
    )
    : null;
  rows.push({
    file: item.file,
    quad: result.quad.map(([x, y]) => [Number(x.toFixed(4)), Number(y.toFixed(4))]),
    meanCornerError: errors ? errors.reduce((sum, value) => sum + value, 0) / errors.length : null,
    maxCornerError: errors ? Math.max(...errors) : null,
    quadIou: errors ? quadIoU(result.quad, item.quad) : null,
    confidence: result.confidence,
    needsReview: result.needsReview,
    method: result.method,
    reviewReasons: result.reviewReasons,
    candidateMethods: result.diagnostics.rankedCandidates?.map((candidate) => candidate.method) ?? [],
    candidateCount: result.diagnostics.candidateCountAfterValidation ?? 0,
    bestScore: result.bestScore,
    secondBestScore: result.secondBestScore,
    selectedFeatures: result.diagnostics.selectedFeatures ?? {},
    selectedWarnings: result.diagnostics.selectedWarnings ?? [],
    confidenceBreakdown: result.diagnostics.confidenceBreakdown,
  });
}

const mean = (values) => values.reduce((sum, value) => sum + value, 0) / Math.max(1, values.length);
const accuracyRows = rows.filter((row) => row.maxCornerError !== null);
const meanErrors = accuracyRows.map((row) => row.meanCornerError).sort((a, b) => a - b);
const metrics = {
  fixture_count: rows.length,
  mean_corner_error: mean(meanErrors),
  median_corner_error: percentile(meanErrors, 0.5),
  max_corner_error_mean: mean(accuracyRows.map((row) => row.maxCornerError)),
  quad_iou_mean: mean(accuracyRows.map((row) => row.quadIou)),
  all_corners_under_1_percent: mean(accuracyRows.map((row) => Number(row.maxCornerError < 0.01))),
  all_corners_under_2_percent: mean(accuracyRows.map((row) => Number(row.maxCornerError < 0.02))),
  review_rate: mean(rows.map((row) => Number(row.needsReview))),
  high_confidence_failure_rate: mean(
    accuracyRows.map((row) => Number(row.confidence >= 0.8 && row.quadIou < 0.75)),
  ),
  runtime_p50_ms: percentile(runtimes, 0.5),
  runtime_p95_ms: percentile(runtimes, 0.95),
};

const payload = `${JSON.stringify(metrics, null, 2)}\n`;
if (outputPath) await writeFile(outputPath, payload);
if (predictionsPath) {
  await writeFile(predictionsPath, `${JSON.stringify({ images: rows }, null, 2)}\n`);
}
process.stdout.write(payload);
